Celery metrics updates

dannyroberts opened the pull request:
Easiest to review commit by commit. The diff looks bigger than it is because two of the commits just move code.

dannyroberts committed:
Fix race condition in time-to-start metric

by recording the start time with before_task_publish instead of after_task_publish

dannyroberts committed:
Move global celery hooks to corehq.celery_monitoring.signals

dannyroberts committed:
Make CELERY_HEARTBEAT_THRESHOLDS overrideable in localsettings

(by moving above localsettings import in settings.py

dannyroberts commented on corehq/celery_monitoring/app_config.py:
```diff
@@ -0,0 +1,11 @@
+
+    def ready(self):
+        # Make sure hooks get registered
+        from .signals import *  # noqa
```
The point of this is to make sure that signals.py gets loaded on startup. Is this the right way to do that? (And maybe less importantly if we decide this is right, what's the way to make stickler ignore it? I thought `# noqa` was it.)

nickpell replied:
I'm not sure if we actually have a way to silence stickler here.  I think including ```# noqa``` as a signal to developers to intentionally ignore sticker comments is the best we can do now.

dannyroberts replied:
Figured it out :) I need both `# noqa` for flake8 and `# pylint: disable=<specific-error>` for pylint

snopoke replied:
does `from . import signals` not work?

nickpell commented on corehq/celery_monitoring/signals.py:
```diff
@@ -0,0 +1,53 @@
+import datetime
```
If you want to be my hero, add imports for both ```absolute_import``` and ```unicode_literals``` here.  Otherwise I have to add them during my weekly ```modernize``` run.

nickpell replied:
🙇

dannyroberts committed:
Add simple test for celery time to start timer

dannyroberts committed:
Address stickler comment

dannyroberts committed:
Have TimeToStartTimer.stop_and_pop_timing raise exception, not return None

dannyroberts committed:
Silence a stickler pylint error

dannyroberts committed:
Sort settings.CELERY_HEARTBEAT_THRESHOLDS

millerdev commented on corehq/celery_monitoring/signals.py:
```diff
@@ -26,12 +30,13 @@ def stop_and_pop_timing(self):
-            cache.delete(self._cache_key)
+        time_sent = cache.get(self._cache_key)
+        if time_sent is None:
+            raise TimingNotAvailable()
```
Did you forget to handle this error below?

https://github.com/dimagi/commcare-hq/blob/d6ca42a4b5e883da2c1b411cb56430304a601596/corehq/celery_monitoring/signals.py#L52

dannyroberts replied:
Oh wow I sure did @millerdev, thanks

dannyroberts replied:
Fixed in new commit

dannyroberts committed:
Fix collection of commcare.celery.heartbeat.blockage_duration metric

it was failing to send because the value passed in was a timedelta instead of a number of seconds

dannyroberts committed:
Fix celery time-to-start collection to catch TimingNotAvailable

forgot to change this when changing TimeToStartTimer.stop_and_pop_timing to raise it.

dannyroberts committed:
Favor non-star import for getting a file run on startup

dannyroberts committed:
For scheduled/retried tasks use task eta as start time

continue using utcnow for normal/non-scheduled tasks, ones without an eta

dannyroberts commented:
Since I'm eager to get this in to today's deploy and the travis build is unlikely to run in this lifetime, I followed emord's suggestion to run the build on another branch: https://github.com/dannyroberts/commcare-hq/pull/1. Assuming the last job on that passes in ~40 minutes, I'm going to treat that as the build passing on this PR

dannyroberts commented:
Build Passed: https://github.com/dannyroberts/commcare-hq/pull/1
